### PR TITLE
[luci] Introduce SubstitutePadV2ToPadPass

### PR DESCRIPTION
**Pull request by hyunsik-yoon**
This introduces `SubstitutePadV2ToPadPass` pass that converts `PadV2` op under a certain condition to `Pad` op.

For #7477
Draft #7478 

Signed-off-by: Hyun Sik Yoon <hyunsik.yoon.1024@gmail.com>

### CODE REVIEW

**Comment by mhs4670go** [comment]
> opt/jenkins_slave/workspace/nnfw/master/pr-nncc-debug/compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp:142:46: error: unused parameter ‘node’ [-Werror=unused-parameter]
14:33:30  bool used_by_maxpool_only(luci::CirclePadV2 *node)
14:33:30                            ~~~~~~~~~~~~~~~~~~~^~~~

**Comment by mhs4670go** [comment]
I think test file is not included mistakenly. Is it intended?

File: compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+ *                 |
+ *         [CircleMaxPool2D]
+ *                 |
+ *
+ * Case 2) Model in Pytorch can be converted to circle (via Onnx)
+ * In such case, it is common that seme 'Reshaping op', e.g., CircleTranspose,
+ * are inserted in-between operations. This pass also needs to handle such situation.
+ *
+ * BEFORE
+ *
```
Comment (mhs4670go) [inline compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp:265]:
```suggestion
```

File: compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+ *          |             |      |            |              |
+ *          -------+-------      -------------+--------------+
+ *                 |                          |
+ *            [CirclePad]                [CirclePadV2]
+ *                 |
+ *         [CircleMaxPool2D]
+ *                 |
+ *
+ * Case 2) Model in Pytorch can be converted to circle (via Onnx)
+ * In such case, it is common that seme 'Reshaping op', e.g., CircleTranspose,
```
Comment (mhs4670go) [inline compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp:261]: seme?
```suggestion
 * In such case, it is common that 'Reshaping op', e.g., CircleTranspose,
```

**Comment by hyunsik-yoon** [comment]
> I think test file is not included mistakenly. Is it intended?

(The tests are in draft) I did not add the implementation of `bool used_by_maxpool_only(luci::CirclePadV2 *node)` to make this PR in reasonable size. So this pass as well as tests do not work.

If adding test is mandatory rule in ONE, I will add `used_by_maxpool_only(..)` function into this PR. Please let me know your opinion. :-)

File: compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+#include <vector>
+
+/**
+ * @brief Convert PadV2 op in a certain condition to Pad
+ * @details Convert PadV2 op in a certain condition to Pad if the following conditions are true:
+ *
+ *    C1) For all i, PadV2.input[i] >= 0
+ *    C2) For all c, PadV2.constant_values[c] <= 0
+ *    C3) PadV2 == MaxPool2D.value()
```
Comment (mhs4670go) [inline compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp:30]: This condition doesn't apply in second case below. Note3 could be added.

**Comment by mhs4670go** [comment]
@hyunsik-yoon 
> Please let me know your opinion. :-)

There's no rules with tests. You can merge this PR as it is:)

**Comment by seanshpark** [comment]
> 14:33:30 /opt/jenkins_slave/workspace/nnfw/master/pr-nncc-debug/compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp:142:46: error: unused parameter ‘node’ [-Werror=unused-parameter]
14:33:30  bool used_by_maxpool_only(luci::CirclePadV2 *node)
14:33:30                            ~~~~~~~~~~~~~~~~~~~^~~~

**Comment by hyunsik-yoon** [comment]
@mhs4670go 

>> Please let me know your opinion. :-)
>
> There's no rules with tests. You can merge this PR as it is:)

Thanks for letting me know. I would like to limit the size of this PR (not to get too much comment!! :-)
Let me add test in subsequent PR. Thanks!

File: compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp (hunk 1 above)
Comment (hyunsik-yoon) [reply compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp]: Fixed! Thanks.

File: compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp (hunk 2 above)
Comment (hyunsik-yoon) [reply compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp]: Fixed!

File: compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp (hunk 3 above)
Comment (hyunsik-yoon) [reply compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp:32]: It was difficult to express C3 that aligns with Note 1 and 2.
So I decided to rewrite the comments. PTAL.

**Comment by hyunsik-yoon** [comment]
Comment in code still look not good. Let me revise more.

**Comment by hyunsik-yoon** [comment]
Revised. PTAL. :-)

File: compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+#include "luci/Pass/SubstitutePadV2ToPadPass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+#include <vector>
+
+/**
+ * @brief Convert PadV2 op in a certain condition to Pad
+ * @details Convert PadV2 op in a certain condition to Pad.
```
Comment (seanshpark) [inline compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp:26]: These two lines are same. Maybe revise `@details` as like "Condition to convert to Pad are like below;`

Comment (hyunsik-yoon) [reply compiler/luci/pass/src/SubstitutePadV2ToPadPass.cpp]: Thanks. Fixed.

**Review by mhs4670go — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGMT thank you

**Pull request merged**